core/src/main/java/com/linecorp/armeria/internal/common/AbstractKeepAliveHandler.java

```diff
@@ -373,6 +460,25 @@ protected void run(ChannelHandlerContext ctx) {
                 // notify the callback.
                 connectionIdleTimeout = executor().schedule(this, connectionIdleTimeNanos,
                                                             TimeUnit.NANOSECONDS);
+
+                if (isActive) {
+                    isActive = false;
+
+                    final ConnectionEventKey connectionEventKey = connectionEventKey(ctx.channel());
+
+                    try {
+                        connectionEventListener.connectionIdle(protocol,
```

Comment:
Idle connections refer to connections that do not currently have any requests in progress. We've encountered a limitation with `AbstractKeepAliveHandler`: it resets the connection's idle time whenever it receives a ping.
https://github.com/line/armeria/blob/e99fa5d535611cc51f7806d6cfecced13abb6029/core/src/main/java/com/linecorp/armeria/internal/common/AbstractKeepAliveHandler.java#L269-L271

Given this behavior, we're considering an alternative approach: utilizing the `unfinishedResponses` property in `AbstractHttpResponseDecoder`. By monitoring changes in `unfinishedResponses`, such as transitions from 0 to 1 indicating activity and from 1 to 0 suggesting idle status, we can effectively manage connection states without interfering with ping handling.
Side note: For HTTP/1.1 connections, the connectionActive and connectionIdle methods will be invoked frequently.

Also, what do you think of splitting this feature into separate PRs for client and server-side?

Reply:
Thank you for the detailed explanation! 🙏 I'll try to utilize `unfinishedResponses`.

> Side note: For HTTP/1.1 connections, the connectionActive and connectionIdle methods will be invoked frequently.

Agreed. Also, I'm a little bit worried about performance when a user adds heavy logic to the handler 🤔

> Also, what do you think of splitting this feature into separate PRs for client and server-side?

Okay. I'll focus on the client side in this PR and apply it to the server side later.

Reply:
> Also, I'm a little bit worried about performance when a user adds heavy logic to the handler 🤔

It shouldn't be a problem because the default is no-op. 😉

Reply:
I'd like to gather idle and keep-alive related logic into `KeepAliveHandler`. It is doable to track idle connections in `AbstractHttpResponseDecoder` but the logic could be fragmented. 

We already call `KeepAliveHandler` when a new request is made.
https://github.com/line/armeria/blob/2219de19d510da43b7fc4afcabf730f8a4cca2ce/core/src/main/java/com/linecorp/armeria/client/AbstractHttpResponseDecoder.java#L67
How about adding `keepAliveHandler().decreasingNumRequests()` and just calling it in `AbstractHttpResponseDecoder.removeResponse()` instead?